File: apps/production-pipeline/packages/agents/src/pre-production/quality-evaluator.ts
import { BaseAgent } from "../base/agent.js";
import type { AgentInput, AgentOutput } from "../base/agent.js";
import type { AIGateway } from "@marionette/ai-gateway";
import type { PrismaClient } from "@prisma/client";
import { RiskAuditor } from "@marionette/analysis-core";

export interface QualityEvaluatorInput extends AgentInput {
  assetPath?: string;
  visualDna?: string;
  setConcept?: string;
  analysisData?: any;
  screenplayFeatures?: any;
}

/**
 * QualityEvaluatorAgent (Industrialized Version)
 * Acts as the final gatekeeper for production assets and creative integrity.
 * Integrates SAIL (Visual SOQ) and RiskAuditor (Quant Intelligence).
 */
export class QualityEvaluatorAgent extends BaseAgent {
  private riskAuditor: RiskAuditor;

  constructor(gateway: AIGateway, db: PrismaClient) {
    super(gateway, db);
    // Note: In a real monorepo setup, you'd inject the predictor correctly.
    // For now, we assume RiskAuditor can be initialized or passed.
    this.riskAuditor = new (RiskAuditor as any)(); 
  }

  async execute(input: QualityEvaluatorInput): Promise<AgentOutput> {
    const { projectId, runId, assetId, assetPath, visualDna, analysisData } = input;

    this.log(`Evaluating integrity for Project: ${projectId} (Asset: ${assetId || "New Concept"})`);
    await this.updateProgress(runId, 20);

    // 1. Creative & Commercial Risk Audit (Level 4 Intelligence)
    let riskReport = { divergenceIndex: 0.1, commercialScore: 85 }; 
    
    try {
      if (analysisData) {
        // In the industrialized pipeline, we use the RiskAuditor from analysis-core
        const audit = await this.riskAuditor.audit(analysisData, input.screenplayFeatures || {});
        riskReport = {
          divergenceIndex: audit.divergenceIndex,
          commercialScore: audit.commercialScore
        };
      }
    } catch (err) {
      this.log(`[Warning] RiskAuditor failed, using safe baseline: ${err}`);
    }

    await this.updateProgress(runId, 50);

    // 2. Visual Quality Assessment (SAIL System Logic)
    const prompt = `
      You are the Senior Quality Assurance Director (SOQ System).
      Evaluate the integrity of the following cinematic asset/concept and provide a SOQ score.
      
      [CONTEXT]
      - Project: ${projectId}
      - Visual DNA: ${visualDna || "Standard Cinematic"}
      - Current Divergence Index: ${riskReport.divergenceIndex}
      
      [REQUIREMENTS]
      - Return JSON only.
      - Fields: 
          "score": (0-100) The Quality/SOQ score.
          "divergenceIndex": Combined with visual feel.
          "decision": "Approved" | "Revision_Required"
          "feedback": Detailed improvement guide for the producing agent.
    `;

    const result = await this.gateway.text(prompt, {
      provider: "gemini" 
    });

    const evalData = JSON.parse(result);
    await this.updateProgress(runId, 80);

    // 3. Persist Evaluation to Asset or PipelineRun
    if (assetId) {
      await this.db.asset.update({
        where: { id: assetId },
        data: {
          metadata: {
            soq_score: evalData.score,
            soq_feedback: evalData.feedback,
            soq_decision: evalData.decision,
            risk_report: riskReport,
            evaluated_at: new Date().toISOString()
          }
        }
      });
      this.log(`Asset ${assetId} SOQ Score: ${evalData.score} | Decision: ${evalData.decision}`);
    }

    await this.updateProgress(runId, 100);

    return {
      success: true,
      message: `Quality evaluation completed. Decision: ${evalData.decision}`,
      data: {
        score: evalData.score ?? 0,
        divergenceIndex: evalData.divergenceIndex ?? riskReport.divergenceIndex,
        decision: evalData.decision,
        feedback: evalData.feedback,
        risk_report: riskReport,
        timestamp: new Date().toISOString()
      }
    };
  }
}

File: apps/production-pipeline/src/agents/quality_evaluator.py
"""
마리오네트 스튜디오 — Quality Evaluator 에이전트 (SAIL System)
표준화된 VisionEngine 인터페이스를 통해 비주얼 에셋의 품질 검증 및 SAIL 무결성 점수(SOQ)를 산출합니다.
"""
import os
import json
from typing import Optional, Dict, Any, List
from dotenv import load_dotenv
from src.engines.factory import EngineFactory


EVALUATOR_SYSTEM_PROMPT = """You are the Senior Quality Assurance Director. 
Evaluate SOQ (Shot Assessment & Integrity Loop) score.
Return JSON: {
  "soq_score": integer,
  "metrics": {"groundedness": int, "aesthetic_alignment": int, "technical_quality": int},
  "feedback": "Korean feedback",
  "actionable_feedback": "English technical directives",
  "decision": "Approved/Revision_Required"
}
"""

class QualityEvaluatorAgent:
    def __init__(self, api_key: str = None):
        load_dotenv(os.path.join(os.path.dirname(__file__), '..', '..', '.env'))
        # UEI: 엔진 팩토리를 통해 비전 엔진 획득
        self.vision_engine = EngineFactory.get_vision_engine()
        print(f"🔍 QualityEvaluator: {self.vision_engine.__class__.__name__} 인터페이스 활성화")

    async def evaluate_asset(self, asset_path: str, visual_dna: str, set_concept: str) -> Dict[str, Any]:
        print(f"🔍 SAIL 검증 시작: {asset_path}")
        
        context = {
            "visual_dna": visual_dna,
            "set_concept": set_concept,
            "system_instruction": EVALUATOR_SYSTEM_PROMPT
        }
        
        result = self.vision_engine.evaluate_quality(asset_path, context)
        
        if result.status == "success":
            try:
                # result.data가 JSON 문자열일 경우 파싱
                eval_data = json.loads(result.data)
                print(f"✅ SOQ SCORE: {eval_data.get('soq_score')} ({eval_data.get('decision')})")
                return eval_data
            except:
                return {"soq_score": 0, "decision": "Error", "feedback": "JSON 파싱 오류"}
        
        return {
            "soq_score": 0,
            "decision": "Error",
            "feedback": f"엔진 오류: {result.data}"
        }
